Autocode java renderers

premprakash90 opened the pull request:
@theRobinator please review and let me know if anything needs to be changed.

Commit:
Autocode java renderers

theRobinator commented on language_utils/java.py:
```diff
@@ -0,0 +1,62 @@
+"""
+
+PRIMITIVE_TYPES = ['byte', 'short', 'int', 'long', 'float', 'double', 'boolean', 'char', 'String']
+BUILT_IN_VARS = {'Infinity', 'NaN', 'null', 'undefined', 'true', 'false', 'window', 'document', 'this',  # Values
```
These are not the Java builtins.

theRobinator commented on language_utils/java.py:
```diff
@@ -0,0 +1,62 @@
+        then instance, public then private, and inheritDoc, typed, and non-typed
+    """
+    def sortKey(item):
+        return (not item.const, not item.static, item.type is None, item.has_prop('private'), item.name)
```
item.type can never be None in Java.

theRobinator commented on renderers/java/classrenderer.py:
```diff
@@ -0,0 +1,57 @@
+        indent_result.extend(x.render(cls) for x in utils.sort_fields(cls.fields.values()))
+        indent_result.append("\n")
+
+    from autocode.primitives import Function
```
You should check for a constructor function to be set here. Java has default constructors.

theRobinator commented on renderers/java/enumrenderer.py:
```diff
@@ -0,0 +1,14 @@
+    if enum.value_type is None:
+        raise Exception('Enums must have type')
+
+    enum.add_prop('enum', enum.value_type.name)
```
I don't think this is a legitimate javadoc tag. This means that we can remove the check for enum.value_type above too.

theRobinator commented on renderers/java/fieldrenderer.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def render(field, owner):
+    if field.has_prop('var'):
```
@var isn't a javadoc tag.

theRobinator commented on renderers/java/fieldrenderer.py:
```diff
@@ -0,0 +1,33 @@
+    if field.type is None:
+        raise Exception('Could not determine the type of the field %s' % field.type)
+
+    field.add_prop('var', field.type.name)
```
Same thing here.

theRobinator commented on renderers/java/fieldrenderer.py:
```diff
@@ -0,0 +1,33 @@
+    if field.const:
+        field.remove_all_props('access')
+
+    if field.description is None and len(field.props) == 0:
```
Looks like this if statement is backwards. It's only rendering the comment if there's nothing to render.

theRobinator commented on renderers/java/fieldrenderer.py:
```diff
@@ -0,0 +1,33 @@
+    else:
+        value_str = ''
+
+    if field.const:
```
The final and static versions are missing visibility below.

theRobinator commented on renderers/java/functionrenderer.py:
```diff
@@ -0,0 +1,53 @@
+            raise Exception('Could not determine the type of the param %s' % param.name)
+
+        if func.return_type is None and func.name != owner.name:
+            raise Exception('Java function must have return type except default constructor')
```
Please change this error to say "All Java functions other than constructors must have return types"

theRobinator commented on renderers/java/valuerenderer.py:
```diff
@@ -0,0 +1,55 @@
+        key, val = value.popitem()
+        key_type = type(key)
+        value_type = type(val)
+        result.append('     put(%s,%s);' % (key, render(val)))
```
In this and on line 40 you need to render the key as well.

theRobinator commented:
It looks like you're also missing an enum for java in languages.py.

Commit:
Corrections from the review.

premprakash90 commented:
@theRobinator please review the fixes. I am not sure how to identify the built in vars in java. I added the built in classes and exception.

Commit:
Changes includes:

- Making enum extend Class instead of Definable
- Classes can contain enum is java, added support for that
- Modified java enum renderer and class renderer
- Modified doc renderer to not render enum twice

Commit:
Removing visibility from Var()

Commit:
- Corrections for the previous review
- Added support for rendering annotations
- Added a new setting render_descriptionless_comments

Commit:
More corrections

Commit:
Annotation renderer does not apply to closure

Commit:
Raise exception when annotationrenderers are called in JS

theRobinator commented:
The latest version of this crashes when generating v5/test/object/get. Also, the PHP layer crashes in the propertyrenderer because EMPTY_COMMENT is not defined.

Run this to generate all the endpoints in java (change the end to "full" or "php" to see the other crash):
./bloom list api | head -21 | tail -20 | sed 's/^ *//' | xargs -n 1 ./bloom java
